src/Evolution/Systems/NewtonianEuler/FiniteDifference/ReconstructWork.tpp

```diff
@@ -0,0 +1,251 @@
+    const Element<Dim>& element,
+    const FixedHashMap<maximum_number_of_neighbors(Dim) + 1,
+                       std::pair<Direction<Dim>, ElementId<Dim>>,
+                       evolution::dg::subcell::NeighborData,
+                       boost::hash<std::pair<Direction<Dim>, ElementId<Dim>>>>
+        neighbor_data,
+    const Mesh<Dim>& subcell_mesh, const size_t ghost_zone_size) noexcept {
+  // Conservative vars tags
+  using MassDensityCons = Tags::MassDensityCons;
+  using EnergyDensity = Tags::EnergyDensity;
+  using MomentumDensity = Tags::MomentumDensity<Dim>;
+
+  // Primitive vars tags
+  using MassDensity = Tags::MassDensity<DataVector>;
+  using Velocity = Tags::Velocity<DataVector, Dim>;
+  using SpecificInternalEnergy = Tags::SpecificInternalEnergy<DataVector>;
+  using Pressure = Tags::Pressure<DataVector>;
+
+  using prim_tags_for_reconstruction =
+      tmpl::list<MassDensity, Velocity, Pressure>;
+
+  ASSERT(Mesh<Dim>(subcell_mesh.extents(0), subcell_mesh.basis(0),
+                   subcell_mesh.quadrature(0)) == subcell_mesh,
+         "The subcell mesh should be isotropic but got " << subcell_mesh);
+  const size_t volume_num_pts = subcell_mesh.number_of_grid_points();
+  const size_t reconstructed_num_pts =
+      (subcell_mesh.extents(0) + 1) *
+      subcell_mesh.extents().slice_away(0).product();
+  const size_t neighbor_num_pts =
+      ghost_zone_size * subcell_mesh.extents().slice_away(0).product();
+  size_t vars_in_neighbor_count = 0;
+  tmpl::for_each<prim_tags_for_reconstruction>(
+      [&element, &neighbor_data, neighbor_num_pts, &reconstruct,
+       reconstructed_num_pts, volume_num_pts, &volume_prims,
+       &vars_in_neighbor_count, &vars_on_lower_face, &vars_on_upper_face,
+       &subcell_mesh](auto tag_v) noexcept {
+        using tag = tmpl::type_from<decltype(tag_v)>;
+        auto& volume_tensor = get<tag>(volume_prims);
+
+        const size_t number_of_components = volume_tensor.size();
+        const gsl::span<const double> volume_vars = gsl::make_span(
+            volume_tensor[0].data(), number_of_components * volume_num_pts);
+        std::array<gsl::span<double>, Dim> upper_face_vars{};
+        std::array<gsl::span<double>, Dim> lower_face_vars{};
+        for (size_t i = 0; i < Dim; ++i) {
+          gsl::at(upper_face_vars, i) = gsl::make_span(
+              get<tag>(gsl::at(*vars_on_upper_face, i))[0].data(),
+              number_of_components * reconstructed_num_pts);
+          gsl::at(lower_face_vars, i) = gsl::make_span(
+              get<tag>(gsl::at(*vars_on_lower_face, i))[0].data(),
+              number_of_components * reconstructed_num_pts);
+        }
+
+        DirectionMap<Dim, gsl::span<const double>> ghost_cell_vars{};
+        for (const auto& direction : Direction<Dim>::all_directions()) {
+          const auto& neighbors_in_direction =
+              element.neighbors().at(direction);
+          ASSERT(neighbors_in_direction.size() == 1,
```

Comment:
why doesn't this fail on external boundaries?

Reply:
Sorry, I'm not sure why you're asking this/why you've concluded it doesn't fail on external boundaries. I'd need to know what code path leads to this passing on an external boundary to investigate and answer the question

Reply:
you are looping over all directions and then accessing the neighbors in each direction...

Reply:
Ah, currently none of the subcell code supports doing subcell at boundaries. The way this is typically handled in FV codes (from what I can understand in the papers) is that ghost cells outside the cell are populated. Specifically, `stencil_width/2+1` are populated so that reconstruction to the external face is also done. Does that answer the question?

Reply:
is no subcell at boundaries enforced (or checked) elsewhere?  I ask because here, line 94 will throw `std::out_of_range` if there is no neighbor which may not be the most helpful error message...

Reply:
It is checked elsewhere. I think this code can support external boundaries, but it is up to the calling code to decide how to populate the external boundary ghost cells.